include/circt/Dialect/HW/HWTypes.h

```diff
@@ -20,6 +20,34 @@
 namespace circt {
 namespace hw {
 class TypedeclOp;
+class TypeAliasType;
+
+template <typename BaseTy>
+class TypeAliasOr
```

Comment:
This approach is neat. I'd like @lattner to weigh in on what he thinks... I'm concerned that there are caveats/subtleties  involved that could cause this to fail in some cases.

Reply:
Also, if this is appropriate there are advantages to having this declared in ODS. Non-parameterized Typedefs should be supported...

Reply:
@teqdruid this is sort of declared in ODS. The ODS TypeAliasOr type is backed by this C++ TypeAliasOr. Did you mean to declare it as a TypeDef instead of a DialectType? I don't think that's desirable for the reason [mentioned](https://github.com/llvm/circt/pull/1143#discussion_r638900259) above.

I'm not sure what you mean about supporting non-parameterized Typedefs... We already support TypeAliasType in most of the HW dialect types (HWValueType, etc.), we just need to support the cases where an op like array_get previously expected just an ArrayType, and we now want to also support aliases there.

Reply:
Typedefs in ODS don't necessarily involve printing/parsing/wrapping. I _think_ it could be used to generate what you've got, but with the additional advantage of being able to use it instead of the type constraints you're writing by hand. I'm imagining a tblgen `class` (which extends typedef) then `def`s for array, struct, and union.

Except for the `classof` override, which is the whole point... Though you could put that in the `extraDeclarations` section.

I'm not particularly opposed to this C++ implementation, though that `classof` override may have implications of which we are not aware.

Reply:
> (which extends typedef)

This is what I initially tried [yesterday](https://github.com/llvm/circt/pull/1143#discussion_r638900259) before deciding a `DialectType` with a custom backing C++ class was more appropriate. I can take another look, but I'm still not sure what would be gained.

As it stands, the ODS `TypeAliasOrConstraint` and `TypeAliasOr` class work together with the C++ `TypeAliasOr` class to transparently let everything (e.g. the parsers, where `parseType` still just works) work with either a `TypeAliasType` or some other type. In other words, there is no way to represent `TypeAliasOr` in the IR, which I think is exactly what we want, and why I chose a `DialectType` rather than a `TypeDef`. My understanding is you would use a TypeDef if you want to have a new type that could be represented in the IR, but that's not really what we want, we just need a "union type" from the perspective of the verifiers, printers, and parsers. Does that make sense? Is there some benefit to using a `TypeDef` here I'm missing?

> though that classof override may have implications of which we are not aware.

`classof` is the hook to customize `isa`,`cast`,`dyn_cast` functionality: https://github.com/llvm/llvm-project/blob/969eefd98e0f8e485148be61190cc2ef62fb1eca/mlir/include/mlir/IR/Types.h#L228

I just checked the MLIR codebase, and indeed this hook is only called by the various implementations of `isa` (and a couple places in the standard dialect / generated code that should probably just call `isa`). So I override it to create the "union type"-like behavior described above, which seems like a valid use case (and empirically works with all existing verifiers, parsers, and printers). Can you be more specific about your concerns? I'm not sure what other implication there could be.

Reply:
I get what you're doing and why. This just doesn't seem like a standard use case to me, so I'm concerned about unknown unknowns, so I'd feel more comfortable if Chris said it was fine. You might bring it up at today's ODM just to get Chris' blessing on this one small thing.

That said, since it works I have to objections to merging it.

Reply:
Sure thing, understood about "unknown unknowns".